### PR TITLE
Port `do` and `observe` functions from PyMC-Experimental

### LINKED ISSUE #5443: Numerical differences between 3.11.4 and 4.0.0b
## Numerical problems with quadratic approximation

As per @ericmjl (https://github.com/pymc-devs/resources/issues/168#issuecomment-1027905552), pymc 3.11.4 and 4.0.0b are giving different numerical answers to quadratic approximation examples for Chapter 2 in the Rethinking Statistics course. Below is an excerpt from the Chapter 2 example. 

pymc 3.11.4 gives standard deviation of 0.16 (correct). 4.0.0b gives 0.64. No other warnings or errors were apparent.

**Please provide a minimal, self-contained, and reproducible example.**
```python
#!/usr/bin/env python
# coding: utf-8

import numpy as np
import pymc3 as pm

RANDOM_SEED = 8927
np.random.seed(RANDOM_SEED)

# #### Code 2.6
# 
# Computing the posterior using the quadratic approximation (quad).

data = np.repeat((0, 1), (3, 6))
with pm.Model() as normal_approximation:
    p = pm.Uniform("p", 0, 1)  # uniform priors
    w = pm.Binomial("w", n=len(data), p=p, observed=data.sum())  # binomial likelihood
    mean_q = pm.find_MAP()
    std_q = ((1 / pm.find_hessian(mean_q, vars=[p])) ** 0.5)[0]

# display summary of quadratic approximation
print("  Mean, Standard deviation\np {:.2}, {:.2}".format(mean_q["p"], std_q[0]))
```

Expected Output

```python
  Mean, Standard deviation
p 0.67, 0.16
```

Actual Output
```python
  Mean, Standard deviation
p 0.67, 0.64
```
</details>

**Please provide any additional information below.**

## Versions and main components

* PyMC/PyMC3 Version: 4.0.0b
* Aesara/Theano Version:
* Python Version: 3.9.7
* Operating system: MacOS
* How did you install PyMC/PyMC3: conda (using environments.yml in rethinking_2)

### PR DESCRIPTION
**Pull request by juanitorduz**
Related to https://github.com/pymc-devs/pymc-experimental/issues/237
Closes pymc-devs/pymc#5443

----
:books: Documentation preview :books:: https://pymc--6879.org.readthedocs.build/en/6879/

### CODE REVIEW

**Comment by ricardoV94** [comment]
I removed the "closes", because we will need a PR that deprecates the functionality it in pymc_experimental in favor of pymc

**Comment by ricardoV94** [comment]
Let's do the following: rename `model/model.py` to `model/basic.py`. Keep all public variables that were present in `model.py` available in `model.__init__.py`

**Comment by twiecki** [comment]
How about `model/core.py` (formerly model.py), `model/transform.py` and `model/fgraph.py`?

**Comment by juanitorduz** [comment]
> How about `model/core.py` (formerly model.py), `model/transform.py` and `model/fgraph.py`?

I am easy! Just provide your feedback and I will change it accordingly :)

**Comment by ricardoV94** [comment]
`transform` should be a submodule. We want to implement many things, and the file would become messy quickly

**Comment by twiecki** [comment]
> `transform` should be a submodule. We want to implement many things, and the file would become messy quickly

Sure, just `model/transform/` then

**Comment by ricardoV94** [comment]
A good check that you are not breaking anything is to not change any of imports in tests (things should be importable from the same "path" after your PR).

**Comment by juanitorduz** [comment]
@ricardoV94, while I fix the `mypy` errors (yay!)... I would appreciate feedback on the new module structure (all tests are passing now). Thanks :)

File: pymc/model/fgraph.py [hunk 1]
```diff
@@ -0,0 +1,402 @@
...
+    return ModelFreeRV(transform=transform)(rv, value, *dims)
+
+
+model_observed_rv = ModelObservedRV()
+model_potential = ModelPotential()
+model_deterministic = ModelDeterministic()
+model_named = ModelNamed()
+
+
+def toposort_replace(
```
Comment (ricardoV94) [inline pymc/model/fgraph.py]: This should go into `pytensorf`

**Comment by ricardoV94** [comment]
Just a small refactor suggestion, otherwise looks great!

**Comment by ricardoV94** [comment]
Ah we also need to list the new functionality in the docs! (And make sure that we show the old model stuff now that the paths have changed)

**Comment by ricardoV94** [comment]
I don't know about a template PR. The goal is for the functionality to be discoverable in the API docs: https://pymcio--6879.org.readthedocs.build/projects/docs/en/6879/api.html

Just need to add some entries in the docs indexes, somewhere here: https://github.com/pymc-devs/pymc/tree/main/docs/source/api

This may be useful: https://www.pymc.io/projects/docs/en/stable/contributing/build_docs.html

File: pymc/model/fgraph.py (hunk 1 above)
Comment (juanitorduz) [reply pymc/model/fgraph.py]: fixed!

**Comment by juanitorduz** [comment]
@ricardoV94 I tried changing the structure in the docs but I am not seeing this locally (see https://github.com/pymc-devs/pymc/pull/6879/commits/adde6d2f11764caaf88a20b40395c9a8e6b0fa0b). I might need some support here 🙏

**Comment by juanitorduz** [comment]
After https://github.com/pymc-devs/pymc/pull/6879/commits/4d7b805674e95a395587b5794da285f207bdd9a8, I can see the docs https://pymcio--6879.org.readthedocs.build/projects/docs/en/6879/api/model/core.html but not in the API page.

**Comment by juanitorduz** [comment]
Ok! I think https://github.com/pymc-devs/pymc/pull/6879/commits/96e6c2bb29f84bd878b4648da0baaa0a4fc1c27b fixed the API entry page.

**Comment by twiecki** [comment]
Just making sure we're making this available from the global namespace so that we can do `pm.do` and `pm.observe`?

**Comment by juanitorduz** [comment]
> Just making sure we're making this available from the global namespace so that we can do `pm.do` and `pm.observe`?

They should be available after https://github.com/pymc-devs/pymc/pull/6879/commits/2fa700bbda690beeb0c3dd34e92c92ed1c1903a5

File: pymc/pytensorf.py [hunk 2]
```diff
@@ -1234,3 +1234,68 @@ def rewrite_pregrad(graph):
...
+class StringConstant(Constant):
+    pass
+
+
+@pytensor._as_symbolic.register(str)
+def as_symbolic_string(x, **kwargs):
+    return StringConstant(stringtype, x)
+
+
+def rvs_in_graph(vars: Sequence[Variable]) -> bool:
```
Comment (ricardoV94) [inline pymc/pytensorf.py]: Where is this function used?

File: pymc/__init__.py [hunk 3]
```diff
@@ -63,7 +63,8 @@ def __set_compiler_flags():
     logsumexp,
     probit,
 )
-from pymc.model import *
+from pymc.model.core import *
+from pymc.model.transform.conditioning import *
```
Comment (ricardoV94) [inline pymc/__init__.py]: We only want `do` and `observe` at the root level. There are other functions there that shouldn't be imported

File: docs/source/api/model/transform/conditioning.rst [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+Others
+------
+
+.. autosummary::
+   :toctree: generated/
+
+   change_value_transforms
+   remove_value_transforms
+   replace_vars_in_graphs
```
Comment (ricardoV94) [inline docs/source/api/model/transform/conditioning.rst]: `replace_vars_in_graph` should not be listed

File: docs/source/api/model/transform/conditioning.rst [hunk 5]
```diff
@@ -0,0 +1,19 @@
+Model Conditioning
```
Comment (ricardoV94) [inline docs/source/api/model/transform/conditioning.rst:1]: Can you add `fgraph_from_model`, `model_from_fgraph` and `clone_model` to the docs as well?

File: pymc/model/transform/conditioning.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+            intervention.name = f"do_{model_var.name}"
+            warnings.warn(
+                f"Intervention expression references the variable that is being intervened: {model_var.name}. "
+                f"Intervention will be given the name: {intervention.name}"
+            )
+        else:
+            intervention.name = model_var.name
+        dims = extract_dims(model_var)
+        # If there are any RVs in the graph we introduce the intervention as a deterministic
+        if rvs_in_graph([intervention]):
```
Comment (juanitorduz) [inline pymc/model/transform/conditioning.py:247]: @ricardoV94 Here we use `rvs_in_graph`

File: docs/source/api/model/transform/conditioning.rst (hunk 5 above)
Comment (juanitorduz) [reply docs/source/api/model/transform/conditioning.rst:1]: fixed in https://github.com/pymc-devs/pymc/pull/6879/commits/aaa91a8e6d14f6a37f5b6ab5ae95043fff44f25e

File: docs/source/api/model/transform/conditioning.rst (hunk 4 above)
Comment (juanitorduz) [reply docs/source/api/model/transform/conditioning.rst]: fixed in https://github.com/pymc-devs/pymc/pull/6879/commits/aaa91a8e6d14f6a37f5b6ab5ae95043fff44f25e

File: pymc/__init__.py (hunk 3 above)
Comment (juanitorduz) [reply pymc/__init__.py]: fixed in https://github.com/pymc-devs/pymc/pull/6879/commits/d669b46d3c04b4e57b73a058bb8c890ad6254888

File: pymc/pytensorf.py (hunk 2 above)
Comment (juanitorduz) [reply pymc/pytensorf.py]: In `pymc/model/transform/conditioning.py` Line 255 if I am not mistaken.

Comment (ricardoV94) [reply pymc/pytensorf.py]: Let's keep it there for now. Not sure yet if it's duplicated functionality or not.

Comment (ricardoV94) [reply pymc/pytensorf.py]: I meant to move this to the file where it's used. I'll do it now

**Comment by ricardoV94** [comment]
I pushed the following changes:
* Reverted changes in imports, so the PR shows less changes
* Updated docstring examples that still showed pymc_experimental imports
* Added __all__ to the new modules
* Rebased from main

Should be good to go when tests pass

**Review by ricardoV94 — APPROVED** [review]

**Comment by juanitorduz** [comment]
Thank you very much @ricardoV94 😀 !

**Pull request merged**